src/main/java/org/truffleruby/core/rope/RopeNodes.java

```diff
@@ -242,6 +244,140 @@ protected static boolean is7Bit(CodeRange codeRange) {
+                                                         @Cached("create()") BranchProfile nonAsciiStringProfile) {
+            CodeRange codeRange = CR_7BIT;
+
+            for (int i = 0; i < bytes.length; i++) {
+                if (bytes[i] < 0) {
+                    nonAsciiStringProfile.enter();
+                    codeRange = CR_VALID;
+                    break;
+                }
+            }
+
+            return new StringAttributes(bytes.length, codeRange);
+        }
+
+        @Specialization(rewriteOn = NonAsciiCharException.class,
+                guards = { "!isEmpty(bytes)", "!isBinaryString(encoding)", "isAsciiCompatible(encoding)" })
+        StringAttributes calculateAttributesAsciiCompatible(Encoding encoding, byte[] bytes) throws NonAsciiCharException {
+            // Optimistically assume this string consists only of ASCII characters. If a non-ASCII character is found,
+            // fail over to a more generalized search.
+            for (int i = 0; i < bytes.length; i++) {
+                if (bytes[i] < 0) {
+                    throw new NonAsciiCharException();
+                }
+            }
+
+            return new StringAttributes(bytes.length, CR_7BIT);
+        }
+
+        @Specialization(replaces = "calculateAttributesAsciiCompatible",
+                guards = { "!isEmpty(bytes)", "!isBinaryString(encoding)", "isAsciiCompatible(encoding)" })
+        StringAttributes calculateAttributesAsciiCompatibleGeneric(Encoding encoding, byte[] bytes,
+                                                                   @Cached("create()") BranchProfile multiByteCharacterProfile,
+                                                                   @Cached("create()") BranchProfile brokenCodeRangeProfile) {
+            // Taken from StringSupport.strLengthWithCodeRangeAsciiCompatible.
+
+            CodeRange codeRange = CR_7BIT;
+            int characters = 0;
+            int p = 0;
+            final int end = bytes.length;
+
+            while (p < end) {
+                if (Encoding.isAscii(bytes[p])) {
+                    final int multiByteCharacterPosition = StringSupport.searchNonAscii(bytes, p, end);
+                    if (multiByteCharacterPosition == -1) {
+                        return new StringAttributes(characters + (end - p), codeRange);
+                    }
+
+                    characters += multiByteCharacterPosition - p;
+                    p = multiByteCharacterPosition;
+                }
+
+                final int lengthOfCurrentCharacter = StringSupport.preciseLength(encoding, bytes, p, end);
+
+                if (lengthOfCurrentCharacter > 0) {
+                    multiByteCharacterProfile.enter();
+
+                    if (codeRange != CR_BROKEN) {
+                        codeRange = CR_VALID;
+                    }
+
+                    p += lengthOfCurrentCharacter;
+                } else {
+                    brokenCodeRangeProfile.enter();
```

Comment:
A single ConditionProfile would be clearer here since both branches are profiled.